### DISCUSSION TITLE
HardReset returns the variable to an inconsistent creation state

### DISCUSSION THREAD

**Discussion by gadavidd** [chocoteam/choco-solver#1105] [General]
Choco Solver's `hardReset` method resets the domain of the variables to their original state. However, there is inconsistent behavior for variables of type `SetVar` and `IntVar` with an enumerated domain. While real variables ( `RealVar` ) and bounded `IntVar` are correctly reset, this is not the case for `SetVar` and enumerated `IntVar` variables.
 
The inconsistency occurs as follows:
 
`SetVar` is reset to `{}` (empty set), regardless of the original domain.
Enumerated `IntVar` directly assumes the first value of the domain instead of maintaining the expected behavior (full original domain).
 
Expected behavior:
 
According to the documentation, `hardReset` should return all variables to the creation state, with their domains complete and unchanged. This behavior works correctly for `RealVar` and bounded `IntVar` variables, but fails for the other configurations mentioned.

```java
Model model = new Model();
model.getEnvironment().worldPush();

SetVar s1 = model.setVar("s1", new int[]{}, new int[]{1, 2, 3, 4, 5});
RealVar r1 = model.realVar("r1", 0, 10, 0.1);
IntVar i1 = model.intVar("i1", 0, 100, true);  // bounded IntVar
IntVar i2 = model.intVar("i2", 0, 100, false); // enumerated IntVar

model.getEnvironment().worldPush();
model.getSolver().hardReset();
```

 For the example, you can see that the domains of `RealVar` (`r1`) and bounded IntVar (`i1`) are correctly restored to the creation state. However, the domains of SetVar (`s1`), which is reset to `{}` (empty set), and Enumerated `IntVar` (` i2` ), which automatically takes on the first value of the domain, do not reflect the expected behavior.
 
On investigation, it seems to me that the problem begins during the creation of the `Environment` type variables for the two cases (`s1` and `i2`), which are either created empty (`SetVar`) or are initially created assuming the first value of the domain (`IntVar`) unlike bounded domain variables. Later, with `hardReset`, this state of inconsistency will be returned, making it impossible to use the mechanism in this context.
 
It seems to me that some treatments could be used to solve this problem:
 
1) Modify the creation of variables in the `Environment`: `SetVar` and enumerated `IntVar` could be initialized with the full domain to avoid inconsistencies;
 
2) Prevent the return to an incorrect state during the `hardReset`.
 
I also tried inserting a new state into the search tree with `worldPush()`, but the behavior remained inconsistent. Are there any suggestions or possible solutions for dealing with this incorrect behavior? It seems to me that only by modifying the shock can this problem be solved.

**Comment by cprudhom** [comment]
I see two options here:
1. Remove the first call to `model.getEnvironment().worldPush()`, 
2. Call `model.getEnvironment().worldCommit()` after the model's population.

**Reply by gadavidd** [reply]
Although the first option solves the problem, removing the first model.getEnvironment().worldPush() is not desirable for the context I'm testing. Although this may seem a bit atypical, it is a necessary behavior, but with the second option I can combine the best of both worlds. 

Thanks again, @cprudhom !
